Improve performance and reduce memory pressure when opening a DICOM file

amoerie opened the pull request:
Fixes #1416
Fixes #1414


#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Replace `ConcurrentBag` with `ConcurrentStack` for the masked DICOM tags in `DicomDictionary`
- This reduces the need for a full in memory snapshot every time we fetch a DICOM tag from the dictionary that is not known and not private
- In .NET Core the memory allocation difference is small but not negligible. In .NET Framework the difference is orders of magnitude higher because `ConcurrentBag.GetEnumerator` is not optimized there yet and it makes multiple intermediate copies.
- Add overload `DicomDataset.AddOrUpdate(DicomItem item)` which avoids an array allocation because you would otherwise hit `DicomDataset.AddOrUpdate(params DicomItem[] item)`. This reduces a lot of allocations made from DicomDatasetReaderObserver, which is responsible for building the DicomDataset instance when reading a file from disk or from the network. 
- Add DicomTagsIndex, a pre-calculated nested dictionary of known DICOM tags by group and element for ultra fast lookup of known DICOM tags. This reduces a lot of allocations made from DicomReader, which parses the bytes of a DICOM file and produces DICOM elements. Before, it would always allocate a new DicomTag instance and then look it up in the dictionary.
- Avoid a closure being created every time we lookup a private creator in `DicomDictionary.GetPrivateCreator`
- Use a precompiled switch table to match incoming bytes against known VRs when parsing an Explicit VR file.

amoerie committed:
Use a concurrent stack to stored masked DICOM tags

amoerie committed:
Update changelog

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1414

amoerie committed:
Add overload to add or update a single DICOM item

This avoids the allocation of an array and two enumerators for every item to add from DicomDatasetReaderObserver when parsing a DICOM file.

amoerie committed:
Don't always allocate a DicomTag

Instead, build an index of known DICOM tags once and use a 2 layer dictionary structure to efficiently look them up

This can only be used for known DICOM tags and not private tags, but it's still worth it.

amoerie committed:
Optimize private creator lookup

- Avoid a lambda + closure for every lookup
- Avoid a method invocation for every creation

amoerie committed:
Optimize entry lookup with 2 layer entry dictionary

Instead of using one dictionary that contains all DICOM tags, use nested structure. The main dictionary contains an entry for every group, and every group entry contains a dictionary of its elements.

amoerie committed:
Undo nested structure in DicomDictionary._entries, no perf difference

amoerie committed:
Update changelog

amoerie commented:
**Performance impact**

We can compare the benchmarks from the `development` branch with this branch to see the effects. Please ignore the execution time, these are not comparable since the benchmarks were executed on different machines. But we _can_ compare the memory allocations.

This is development, from the (unmodified) ParseDatasetBenchmarks:

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.20348
Intel Xeon Platinum 8370C CPU 2.80GHz, 1 CPU, 2 logical and 2 physical cores
.NET SDK=6.0.301
  [Host]     : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT
  DefaultJob : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT


```
|        Method |     Mean |   Error |  StdDev |   Gen 0 |  Gen 1 | Allocated |
|-------------- |---------:|--------:|--------:|--------:|-------:|----------:|
|       ParseCT | 497.4 μs | 1.68 μs | 1.57 μs |  9.7656 | 1.9531 |    251 KB |
|       ParseMR | 576.7 μs | 2.06 μs | 1.93 μs | 12.6953 | 2.9297 |    327 KB |
| ParseDicomdir | 564.6 μs | 1.75 μs | 1.64 μs | 11.7188 | 3.9063 |    309 KB |


And this branch GH-1416: (note that this includes the fixes made for #1414)

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.20348
Intel Xeon Platinum 8370C CPU 2.80GHz, 1 CPU, 2 logical and 2 physical cores
.NET SDK=6.0.301
  [Host]     : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT
  DefaultJob : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT


```
|        Method |     Mean |   Error |  StdDev |  Gen 0 |  Gen 1 | Allocated |
|-------------- |---------:|--------:|--------:|-------:|-------:|----------:|
|       ParseCT | 385.0 μs | 1.55 μs | 1.45 μs | 2.4414 | 0.4883 |     69 KB |
|       ParseMR | 431.4 μs | 1.12 μs | 0.99 μs | 3.4180 | 0.4883 |     84 KB |
| ParseDicomdir | 537.6 μs | 2.37 μs | 1.98 μs | 8.7891 | 2.9297 |    231 KB |


Notice how the parsing went down from 251 KB -> 69KB for a CT and from 327 KB -> 84KB for an MR.

mrbean-bremen commented on FO-DICOM.Core/DicomDictionary.cs:
```diff
@@ -307,14 +307,27 @@ public void Add(DicomDictionaryEntry entry)
         {
-            return _creators.GetOrAdd(creator, _ => new DicomPrivateCreator(creator));
+            while (true)
+            {
```
Why this loop is needed? You would get here if adding the private creator failed, why would that work on a second try? And if it didn't, won't you get an infinite loop here?

amoerie replied:
You should imagine thousands of threads actively trying to add or update the same private creator here. TryAdd can fail if the value was already added between when you checked if it existed in the dictionary and the TryAdd call.
You can not end in an infinite loop because eventually TryAdd or TryGetValue will succeed. This pattern is an alternative to using locking and is usually a lot faster in highly concurrent environments.

mrbean-bremen replied:
Ah right, didn't see that. I know this pattern as "double check" or something like that - haven't seen it in a loop, but it makes sense.

mrbean-bremen commented on FO-DICOM.Core/DicomTagsIndex.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace FellowOakDicom
+{
+    public static class DicomTagsIndex
```
It is mostly self-explaining, but maybe add some summary here...

amoerie replied:
Ah good call. Do you think this should be public or should I set it to internal and keep it for performance purposes only?

mrbean-bremen replied:
I think it is sufficient to keep it internal, at the moment I don't see any need to expose it.

amoerie replied:
I've made this class internal, added some documentation. 
Furthermore, the implementation was refactored from a dictionary based approach to a 2D array based approach.
Due to the distribution of group and element values, this 2D array is rather sparse (a lot of empty cells), which causes the full index to consume about 6 MB (allocated exactly once, lazily on demand).
The dictionary approach consumed about 0.4 MB in comparison.
The upside is that lookups against a 2D array are ~6x faster than lookups against a dictionary. This seemed worthwhile to me.

amoerie committed:
Use 2D array for DICOM tags index

amoerie committed:
Make DICOM tags index internal

gofal commented on FO-DICOM.Core/DicomTagsIndex.cs:
```diff
@@ -0,0 +1,46 @@
+{
+    public static class DicomTagsIndex
+    {
+        private static Lazy<Dictionary<ushort, Dictionary<ushort, DicomTag>>> Index = new Lazy<Dictionary<ushort, Dictionary<ushort, DicomTag>>>(BuildIndex);
```
`ushort group` and `ushort element` can be combined into one `uint tag`. Wouldnt one Dictionary<uint, DicomTag> be more efficient?

amoerie replied:
I had the same thought when implementing this. After benchmarking this, these are the numbers:
- The single array is 20x larger than a 2D array (120MB memory usage vs 6 MB memory usage)
- Lookups against the single array are ~15% faster than against the 2D array.

I personally think the 120MB memory usage is too high and not worth it.

Furthermore, my benchmark sequentially accesses every DICOM tag. With random access, the performance is worse due to more cache misses. Since the single array is so large, it is too big for L1 and L2 caches in the processor.

gofal replied:
wow, never thought that the single array is so bad in memory usage

amoerie replied:
The 2D array is "clever" in that it only allocates an array per group of length "max-element-value" which has a compounding effect. 
Since the sequence closing tags are of group + element FFFFF, the single 1D array must be uint.MaxValue long, which is highly inefficient.

amoerie replied:
So, I went a little more in-depth into this implementation. This is the full benchmark + implementation code:

```csharp
// Copyright (c) 2012-2021 fo-dicom contributors.
// Licensed under the Microsoft Public License (MS-PL).

using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using FellowOakDicom;

namespace FellowOakDicom.Benchmark
{
    static class Program
    {
        static void Main()
        {
            // Run all benchmarks in assembly
            BenchmarkRunner.Run<DicomTagsIndexSizeBenchmarks>();
            BenchmarkRunner.Run<DicomTagsIndexLookupBenchmarks>();
        }
    }
}

[MemoryDiagnoser]
public class DicomTagsIndexSizeBenchmarks
{
    [Benchmark(Baseline = true)]
    public void Dictionary_BuildIndex()
    {
        for (int i = 0; i < 100; i++)
        {
            var index = DicomTagsIndexToBenchmark.BuildDictionaryIndex();
            if (index == null)
            {
                throw new Exception("Index is null");
            }
        }
    }

    [Benchmark()]
    public void Dictionary2D_BuildIndex()
    {
        for (int i = 0; i < 100; i++)
        {
            var index = DicomTagsIndexToBenchmark.BuildDictionary2DIndex();
            if (index == null)
            {
                throw new Exception("Index is null");
            }
        }
    }

    [Benchmark]
    public void Array2D_BuildIndex()
    {
        for (int i = 0; i < 100; i++)
        {
            var index = DicomTagsIndexToBenchmark.BuildArray2DIndex();
            if (index == null)
            {
                throw new Exception("Index is null");
            }
        }
    }
}

[MemoryDiagnoser]
public class DicomTagsIndexLookupBenchmarks
{
    private DicomTag[] _allKnownDicomTags;

    [GlobalSetup]
    public void GlobalSetup()
    {
        _allKnownDicomTags = typeof(DicomTag)
            .GetFields(BindingFlags.Public | BindingFlags.Static)
            .Where(field => field.FieldType == typeof(DicomTag))
            .Select(field => field.GetValue(null) as DicomTag)
            .Where(tag => tag != null)
            .ToArray();

        switch (Access)
        {
            case "Sequential":
                break;
            case "Random":
                Shuffle(_allKnownDicomTags);
                break;
            default:
                throw new ArgumentOutOfRangeException();
        }
    }

    public static void Shuffle<T>(IList<T> list)
    {
        var rng = new Random();
        int n = list.Count;
        while (n > 1) {
            n--;
            int k = rng.Next(n + 1);
            (list[k], list[n]) = (list[n], list[k]);
        }
    }

    [Params("Sequential", "Random")]
    public string Access { get; set; }

    private void PerformLookups()
    {
        for (var i = 0; i < _allKnownDicomTags.Length; i++)
        {
            var knownDicomTag = _allKnownDicomTags[i];
            var tag = DicomTagsIndexToBenchmark.Lookup(knownDicomTag.Group, knownDicomTag.Element);
            if (tag == null)
            {
                throw new Exception("Tag not found");
            }
        }
    }

    [Benchmark(Baseline = true)]
    public void Dictionary()
    {
        DicomTagsIndexToBenchmark.Type = DicomTagsIndexToBenchmark.IndexType.Dictionary;
        PerformLookups();
    }

    [Benchmark()]
    public void Dictionary2D()
    {
        DicomTagsIndexToBenchmark.Type = DicomTagsIndexToBenchmark.IndexType.Dictionary2D;
        PerformLookups();
    }

    [Benchmark]
    public void Array2D()
    {
        DicomTagsIndexToBenchmark.Type = DicomTagsIndexToBenchmark.IndexType.Array2D;
        PerformLookups();
    }
}

/// <summary>
/// Prebuilt index for high-performance lookup of known DICOM tags
/// </summary>
public static class DicomTagsIndexToBenchmark
{
    public enum IndexType
    {
        Dictionary2D,
        Dictionary,
        Array2D
    }

    public static IndexType Type = IndexType.Dictionary2D;

    private static Lazy<Dictionary<uint, DicomTag>> DictionaryIndex
        = new Lazy<Dictionary<uint, DicomTag>>(BuildDictionaryIndex);

    private static Lazy<Dictionary<ushort, Dictionary<ushort, DicomTag>>> Dictionary2DIndex
        = new Lazy<Dictionary<ushort, Dictionary<ushort, DicomTag>>>(BuildDictionary2DIndex);

    private static Lazy<DicomTag[][]> Array2DIndex
        = new Lazy<DicomTag[][]>(BuildArray2DIndex);

    private static List<DicomTag> GetAllDicomTags()
    {
        return typeof(DicomTag)
            .GetFields(BindingFlags.Public | BindingFlags.Static)
            .Where(field => field.FieldType == typeof(DicomTag))
            .Select(field => field.GetValue(null) as DicomTag)
            .Where(tag => tag != null)
            .ToList();
    }

    public static Dictionary<uint, DicomTag> BuildDictionaryIndex()
    {
        var dicomTags = GetAllDicomTags();

        var index = new Dictionary<uint, DicomTag>();
        foreach (var tag in dicomTags)
        {
            var key = ((uint)tag.Group << 8) | tag.Element;
            if (!index.ContainsKey(key))
            {
                index[key] = tag;
            }
        }


        return index;
    }

    public static Dictionary<ushort, Dictionary<ushort, DicomTag>> BuildDictionary2DIndex()
    {
        var dicomTags = GetAllDicomTags();
        var dicomTagsPerGroup = dicomTags
            .GroupBy(tag => tag.Group)
            .ToList();

        var index = new Dictionary<ushort, Dictionary<ushort, DicomTag>>(dicomTagsPerGroup.Count);
        foreach (var group in dicomTagsPerGroup)
        {
            var tagsInGroup = group.ToList();
            var groupIndex = new Dictionary<ushort, DicomTag>(tagsInGroup.Count);

            foreach (var tag in group)
            {
                if (!groupIndex.ContainsKey(tag.Element))
                {
                    groupIndex.Add(tag.Element, tag);
                }
            }

            index[group.Key] = groupIndex;
        }

        return index;
    }

    public static DicomTag[][] BuildArray2DIndex()
    {
        var dicomTagsPerGroup = GetAllDicomTags()
            .GroupBy(tag => tag.Group)
            .OrderByDescending(group => group.Key)
            .ToList();

        var highestGroup = dicomTagsPerGroup[0].Key;
        var index = new DicomTag[highestGroup + 1][];

        foreach (var group in dicomTagsPerGroup)
        {
            var tagsInGroup = group.OrderByDescending(t => t.Element).ToList();
            var highestElement = tagsInGroup[0].Element;
            var groupIndex = new DicomTag[highestElement + 1];

            foreach (var tag in group)
            {
                if (groupIndex[tag.Element] == null)
                {
                    groupIndex[tag.Element] = tag;
                }
            }

            index[group.Key] = groupIndex;
        }

        return index;
    }

    public static DicomTag Lookup(ushort group, ushort element)
    {
        switch (Type)
        {
            case IndexType.Dictionary:
            {
                return DictionaryIndex.Value.TryGetValue(((uint)group << 8) | element, out var tag)
                    ? tag
                    : null;
            }
            case IndexType.Dictionary2D:
            {
                return Dictionary2DIndex.Value.TryGetValue(group, out var tagsInGroup)
                    && tagsInGroup.TryGetValue(element, out var tag)
                    ? tag
                    : null;
            }
            case IndexType.Array2D:
                return Array2DIndex.Value[group]?[element];
            default:
                throw new ArgumentOutOfRangeException();
        }
    }

}
```

This produces the following benchmark output:

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1766 (21H2)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.301
  [Host]     : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT
  DefaultJob : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT

**Size**

```
|                  Method |      Mean |    Error |    StdDev | Ratio | RatioSD |       Gen 0 |       Gen 1 |      Gen 2 | Allocated |
|------------------------ |----------:|---------:|----------:|------:|--------:|------------:|------------:|-----------:|----------:|
|   Dictionary_BuildIndex | 110.73 ms | 2.933 ms |  8.509 ms |  1.00 |    0.00 |  18000.0000 |   9000.0000 |  9000.0000 |     80 MB |
| Dictionary2D_BuildIndex |  71.95 ms | 1.353 ms |  2.508 ms |  0.65 |    0.06 |  15875.0000 |   1500.0000 |          - |     53 MB |
|      Array2D_BuildIndex | 375.86 ms | 9.404 ms | 27.727 ms |  3.42 |    0.38 |  61000.0000 |  56000.0000 | 39000.0000 |    728 MB |

__Note__: Since the benchmark builds 100 indexes, you need to divide the allocation by 100. So Dictionary = 0.8MB, Dictionary2 = 0.5MB and Array2D = 7,3MB

**Lookup speed**

|       Method |     Access |      Mean |    Error |    StdDev |    Median | Ratio | RatioSD | Allocated |
|------------- |----------- |----------:|---------:|----------:|----------:|------:|--------:|----------:|
|   **Dictionary** |     **Random** | **104.52 μs** | **2.071 μs** |  **2.835 μs** | **103.06 μs** |  **1.00** |    **0.00** |         **-** |
| Dictionary2D |     Random | 232.23 μs | 5.302 μs | 15.466 μs | 226.86 μs |  2.20 |    0.14 |         - |
|      Array2D |     Random |  35.09 μs | 0.619 μs |  0.687 μs |  35.16 μs |  0.33 |    0.01 |         - |
|              |            |           |          |           |           |       |         |           |
|   **Dictionary** | **Sequential** |  **72.32 μs** | **0.881 μs** |  **0.781 μs** |  **72.35 μs** |  **1.00** |    **0.00** |         **-** |
| Dictionary2D | Sequential | 140.90 μs | 2.670 μs |  2.497 μs | 140.22 μs |  1.95 |    0.04 |         - |
|      Array2D | Sequential |  26.04 μs | 0.662 μs |  1.932 μs |  25.74 μs |  0.36 |    0.03 |         - |


Based on these numbers I would maybe choose for the Dictionary solution after all. Its performance its 3 times worse for lookups, but it requires 10x less memory...

The 120MB I mentioned before was a flat Array by the way, which also suffered from index out of range errors for very large groups + elements, so I've removed this implementation from the investigation.

amoerie replied:
From the failing tests, it looks like there were still opportunities for IndexOutOfRangeExceptions to occur.
For maximum safety, I've switched to the Dictionary approach. Especially for people who use fo-dicom to open a single file, a 7MB allocation would be over the top..

gofal commented on FO-DICOM.Core/DicomTagsIndex.cs:
```diff
@@ -0,0 +1,46 @@
+            return index;
+        }
+
+        internal static DicomTag Lookup(ushort group, ushort element) =>
```
every single call to "DicomTagsIndex.Lookup" now also requires 'DicomTagsIndex.lookkup(group, element) ?? new DicomTag(group, element)`
So for simplicity (and backwardscompatibility), this method LookUp could be changed to "Get" and internally lookup and return in case it was found, and create a new one otherwise (that then maybe is added to the dictionary?)

amoerie replied:
We could incorporate the "?? new DicomTag(group, element)" in the Lookup method, but then it 'violates' the name of the method because it wasn't really found in the index. I'm not sure if it makes sense, but it would make our usages more DRY. 

Also, the index is completely readonly right now. Dictionary is thread safe when it's readonly, but if we will add tags there on-demand then you have to introduce locking or switch to a ConcurrentDictionary, which would have a negative impact on the performance. I would prefer to keep it readonly. We can change the name to "KnownDicomTagsIndex" if that makes it clearer.

Edit: I've renamed the method to "LookupOrCreate", is that okay? I still wouldn't add anything on demand though.

gofal replied:
Hi thanks for the explanation. keeping the dictionary readable is reasonable.

And sure, when changing the behavior of the method and implicitly creating a new DicomTag if is not found in the dictionary, then also the name should be changed accordingly. My suggestion "Get" was too general. "LookupOrCreate" is prety exactly what the method does.

amoerie commented:
These are benchmarks run on my machine, so we can compare the execution time too.

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1766 (21H2)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.301
  [Host]     : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT
  DefaultJob : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT

```

**development**

|        Method |     Mean |    Error |   StdDev |   Median |    Gen 0 |   Gen 1 | Allocated |
|-------------- |---------:|---------:|---------:|---------:|---------:|--------:|----------:|
|       ParseCT | 528.3 μs | 10.44 μs | 25.62 μs | 519.5 μs |  81.0547 |       - |    251 KB |
|       ParseMR | 628.4 μs | 11.58 μs | 10.83 μs | 626.7 μs | 106.4453 |       - |    327 KB |
| ParseDicomdir | 623.7 μs | 12.44 μs | 22.44 μs | 617.8 μs |  75.1953 | 26.3672 |    309 KB |

**GH-1416**

|        Method |     Mean |    Error |   StdDev |   Gen 0 |   Gen 1 | Allocated |
|-------------- |---------:|---------:|---------:|--------:|--------:|----------:|
|       ParseCT | 388.1 μs |  7.74 μs | 12.27 μs | 21.9727 |  1.4648 |     68 KB |
|       ParseMR | 441.5 μs |  8.67 μs | 11.27 μs | 26.8555 |  0.4883 |     84 KB |
| ParseDicomdir | 562.7 μs | 11.27 μs | 32.68 μs | 55.6641 | 18.5547 |    231 KB |

amoerie committed:
Improve DicomTagsIndex

Use a flat dictionary approach. This is slightly slower for lookups but consumes a lot less memory
Also create a DICOM tag instance if it is not found in the known tags index

amoerie committed:
Use the correct amount of bit shifting and add tests for DicomTagsIndex

amoerie committed:
Use precompiled UTF8 VR switch clause

Instead of producing a lot of intermediate strings using an UTF8 encoder, simply parse VRs from their raw byte array which is UTF8 encoded anyway
For the special case where the VR is just empty or white space, we do need to get the UTF8 decoded string to check for that, but now we only materialize that string in the unhappy case where known VR matching failed

mrbean-bremen commented on FO-DICOM.Core/DicomVR.cs:
```diff
@@ -175,6 +175,121 @@ private static DicomVR TryParse(string vr, out bool valid)
+        {
+            valid = true;
+            // This is a precompiled switch table
+            switch (vr[0])
```
Is it ensured here that the length of `vr` is at least 2? At least in the second usage this seems not to be the case,but I maybe wrong.

amoerie replied:
The previous implementation also called GetBytes(2) so that behavior has stayed the same..

mrbean-bremen replied:
Ah, right - I wasn't aware that `GetBytes` throws if it doesn't get all bytes.

mrbean-bremen replied:
Ok, you decided you play it save :)

amoerie committed:
Guard against invalid vr bytes when trying to parse

amoerie committed:
Add IByteSource.GetBytes(byte[] buffer, int index, int count)

amoerie committed:
Avoid a lot of intermediate byte array allocations when parsing VRs

amoerie committed:
Merge branch 'development' into ReadIntoBuffers

amoerie committed:
Merge branch 'development' into GH-1416

amoerie committed:
Implement reading into buffer in UnseekableStreamByteSource

amoerie committed:
Fix check for bad private sequence

amoerie committed:
Merge branch 'ReadIntoBuffers' into GH-1416

amoerie commented:
I've managed to reduce parsing memory usage by another 5kb for Explicit VR files because for VR parsing, GetBytes(int count) was called for each tag, which allocates a 2 byte array. While small, this does add up because it's repeated for each tag.
This was improved by adding a method to IByteSource called GetBytes(byte[] buffer, int index, int count) (a signature that is also available on BinaryReader) and using that overload in several places.
The tricky part was to also implement this in UnseekableStreamByteSource and ByteBufferByteSource, but I think I managed to do so.

Latest numbers:

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22622
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.301
  [Host]     : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT
  DefaultJob : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT


```

**development**

|         Method |     Mean |   Error |  StdDev |   Gen 0 |   Gen 1 | Allocated |
|--------------- |---------:|--------:|--------:|--------:|--------:|----------:|
| CT_LE_Implicit | 356.9 μs | 2.79 μs | 2.61 μs | 40.5273 |  9.7656 |    251 KB |
| MR_LE_Implicit | 427.6 μs | 1.54 μs | 1.44 μs | 53.2227 | 13.1836 |    327 KB |
| MG_LE_Explicit | 152.0 μs | 1.40 μs | 1.31 μs | 15.3809 |  3.1738 |     94 KB |
|       DICOMDIR | 403.5 μs | 2.33 μs | 2.18 μs | 50.2930 | 17.0898 |    309 KB |


**GH-1416**

|         Method |     Mean |   Error |  StdDev |   Gen 0 |   Gen 1 | Allocated |
|--------------- |---------:|--------:|--------:|--------:|--------:|----------:|
| CT_LE_Implicit | 266.3 μs | 1.62 μs | 1.52 μs | 10.7422 |  2.4414 |     68 KB |
| MR_LE_Implicit | 299.6 μs | 1.00 μs | 0.94 μs | 13.1836 |  3.4180 |     84 KB |
| MG_LE_Explicit | 136.8 μs | 0.94 μs | 0.88 μs |  8.5449 |  2.6855 |     54 KB |
|       DICOMDIR | 367.5 μs | 2.88 μs | 2.55 μs | 31.2500 | 15.6250 |    193 KB |

amoerie committed:
Merge branch 'GH-1416' of github.com:fo-dicom/fo-dicom into GH-1416

amoerie committed:
Implement GetBytes(int count) using GetBytes(byte[] buffer, int index, int count)

amoerie committed:
Make GH-1359 test less flaky

One day, we'll find a better way to write these time based tests...

amoerie committed:
Merge branch 'development' into GH-1416

amoerie committed:
Use string as key for private dictionary

amoerie committed:
Explicitly choose ordinal string comparer

(This was already the default value)

amoerie committed:
Initialize DICOM tags index with correct size

amoerie committed:
Merge branch 'GH-1416' of github.com:fo-dicom/fo-dicom into GH-1416

amoerie committed:
Merge branch 'development' into GH-1416

amoerie commented:
@gofal I think your comments are all resolved now, if you would find the time for a second pass it would be much appreciated. Thanks!

gofal reviewed (approved):
Looks great, many thanks!

amoerie committed:
Merge branch 'GH-1416' of github.com:fo-dicom/fo-dicom into GH-1416

amoerie committed:
Merge with development

gofal committed:
Merge branch 'development' into GH-1416
